src/Database/LSMTree/Internal/Snapshot.hs

```diff
@@ -287,23 +168,86 @@ snapIncomingRun (Merging MergingRun {..}) = do
         mergeCompletedCache
         smrs
 
-{-# SPECIALISE snapMergingRunState :: MergingRunState IO h -> IO SnapMergingRunState #-}
+{-# SPECIALISE snapMergingRunState ::
+     MergingRunState IO h
+  -> IO (SnapMergingRunState (Run IO h)) #-}
 snapMergingRunState ::
      PrimMonad m
   => MergingRunState m h
-  -> m SnapMergingRunState
-snapMergingRunState (CompletedMerge r) = pure (SnapCompletedMerge (runNumber r))
+  -> m (SnapMergingRunState (Run m h))
+snapMergingRunState (CompletedMerge r) = pure (SnapCompletedMerge r)
 -- We need to know how many credits were spent already so we can restore merge
 -- work on snapshot load.
 snapMergingRunState (OngoingMerge rs (SpentCreditsVar spentCreditsVar) m) = do
     spentCredits <- readPrimVar spentCreditsVar
-    pure (SnapOngoingMerge (V.map runNumber rs) (SpentCredits spentCredits) (Merge.mergeLevel m))
+    pure (SnapOngoingMerge rs (SpentCredits spentCredits) (Merge.mergeLevel m))
+
+{-------------------------------------------------------------------------------
+  Runs
+-------------------------------------------------------------------------------}
 
-runNumber :: Run m h -> RunNumber
-runNumber r = Paths.runNumber (Run.runRunFsPaths r)
+-- TODO: temp registry
```

Comment:
Do you still want to tackle this here or later?

Reply:
Done!